File: web/src/lib/markdown/textStatistics.ts
export interface TextStats {
	wordCount: number;
	characterCount: number;
	lineCount: number;
}

export function computeTextStatistics(text: string): TextStats {
	const characterCount = text.length;

	if (text.length === 0) {
		return { wordCount: 0, characterCount: 0, lineCount: 0 };
	}

	// Split on whitespace, omitting empty subsequences (matches Swift's split(omittingEmptySubsequences: true))
	const words = text.split(/\s+/).filter((w) => w.length > 0);
	const wordCount = words.length;

	// Count newline characters (\n and \r), matching Swift's isNewline
	const newlineCount = [...text].filter((ch) => ch === '\n' || ch === '\r').length;
	const lineCount = newlineCount + 1;

	return { wordCount, characterCount, lineCount };
}

File: MarkdownNotes/Sources/MarkdownNotes/Utilities/TextStatistics.swift
import Foundation

struct TextStatistics {
    let wordCount: Int
    let characterCount: Int
    let lineCount: Int

    init(from text: String) {
        characterCount = text.count

        if text.isEmpty {
            wordCount = 0
            lineCount = 0
            return
        }

        // Word count: split by whitespace and newlines, filter empty components
        let words = text.split(omittingEmptySubsequences: true) { $0.isWhitespace }
        wordCount = words.count

        // Line count: number of newline characters + 1
        // "hello" = 1, "a\nb" = 2, "hello\n" = 2
        let newlineCount = text.filter { $0.isNewline }.count
        lineCount = newlineCount + 1
    }
}
